File: misc/abacaba.c
/* ABACABA state machine
 * Ref: https://redd.it/njxq95
 * This is free and unencumbered software released into the public domain.
 */

/* Compute the next ABACABA state. The initial state is zero, and halt
 * is indicated by returning to the zero state.
 *
 * The state is a 31-bit quantity where bits 0-24 are a bitstack, bits
 * 25-29 are the stack size, and bit 30 is the recursion direction.
 *
 *     D IIIII SSSSSSSSSSSSSSSSSSSSSSSSS
 */
static long
abacaba(long s)
{
    for (;;) {
        long stack = s & 0x1ffffff;
        int i = s>>25 & 0x1f;
        int descending = s>>30;
        int middle = s>>i & 1;

        if (i == 25) {
            // bottom out, descend to the parent
            return 1L<<30 | (i-1L)<<25 | stack;
        } else if (descending && !middle) {
            // output "middle" character, ascend into right branch
            return (i+1L)<<25 | stack | 1L<<i;
        } else if (descending && middle) {
            if (!i) return 0L; // halt
            // descend to parent
            s = 1L<<30 | (i-1L)<<25 | (stack ^ 1L<<i);
        } else {
            // ascend into left branch
            s = (i+1L)<<25 | stack;
        }
    }
}

/* Return the output letter for a given state. */
static int
abacaba_letter(long s)
{
    return 'z' - (s>>25 & 0x1f) + (s>>30 ? -1: +1);
}

/* Usage example */
#include <stdio.h>

int
main(void)
{
    for (long state = abacaba(0); state; state = abacaba(state)) {
        putchar(abacaba_letter(state));
    }
    putchar('\n');
}

File: misc/abacaba.go
// ABACABA state machine
// Ref: https://redd.it/njxq95
// This is free and unencumbered software released into the public domain.
package main

import (
	"bufio"
	"os"
)

type State uint32

// Iterate the state matching, returning the next letter of the ABACABA
// sequence, the next state, and whether or not the machine has halted.
// The initial state is zero.
//
// The state is a 32-bit quantity where bits 0-25 are a bitstack, bits
// 26-30 are the stack size, and bit 31 is the recursion direction.
//
//     D IIIII SSSSSSSSSSSSSSSSSSSSSSSSSS
func (s State) Next() (rune, State, bool) {
	for {
		stack := s & 0x3ffffff
		i := s >> 26 & 0x1f
		descending := s>>31 == 1
		middle := s>>i&1 == 1

		if i == 25 {
			// Bottom out, descend back to the parent
			return 'a', State(1)<<31 | (i-1)<<26 | stack, false
		} else if descending && !middle {
			// Output "middle" character, ascend into right branch
			return 'z' - rune(i), (i+1)<<26 | stack | State(1)<<i, false
		} else if descending && middle {
			if i == 0 {
				// At root, halt
				return 0, 0, true
			}
			// Descend back to the parent
			s = State(1)<<31 | (i-1)<<26 | stack ^ State(1)<<i
		} else {
			// Ascend into the left branch
			s = (i+1)<<26 | stack
		}
	}
}

func main() {
	buf := bufio.NewWriter(os.Stdout)
	for c, s, done := State(0).Next(); !done; c, s, done = s.Next() {
		buf.WriteRune(c)
	}
	buf.WriteRune('\n')
	buf.Flush()
}
